src/librbd/WatchNotifyTypes.h

```diff
@@ -200,11 +200,12 @@ struct ResizePayload : public AsyncRequestPayloadBase {
   static const NotifyOp NOTIFY_OP = NOTIFY_OP_RESIZE;
   static const bool CHECK_FOR_REFRESH = true;
 
-  ResizePayload() : size(0) {}
-  ResizePayload(uint64_t size_, const AsyncRequestId &id)
-    : AsyncRequestPayloadBase(id), size(size_) {}
+  ResizePayload() : size(0), can_shrink(true) {}
+  ResizePayload(uint64_t size_, bool can_shrink_, const AsyncRequestId &id)
+    : AsyncRequestPayloadBase(id), size(size_), can_shrink(can_shrink_) {}
 
   uint64_t size;
+  bool can_shrink;
```

Comment:
You need to modify the ResizePayload ```encode```, ```decode```, and ```dump``` methods to handle the new param. You will need to up the version from 2 -> 3 and only attempt to decode the ```allow_shrink``` param if version >= 3 (see ```AcquiredLockPayload::decode``` for an example)